File: server/tools/acknowledge_notice.js
// Marks a Heartbeat/Trigger finding (server/heartbeat/) as delivered once
// Jarvis has actually told the owner about it in a live turn — the
// resolving action a source:'heartbeat' outbox row has no equivalent of
// otherwise (a background Job's own tier1/2 rows resolve via
// check_on_work/stop_working_on instead; see prompt.js's jobsSection() for
// the full wording). No confirm gate: recording that something was said has
// no outward effect of its own to protect.
//
// Leaf-safe: imports only heartbeat/outbox-store.js, itself a leaf.

import { getById, markDelivered } from '../heartbeat/outbox-store.js';

export default {
  name: 'acknowledge_notice',
  core: true,
  meta: true,
  description:
    'Call this right after you have actually told the owner about a background notice mentioned in your own instructions ' +
    '(a notice_id shown alongside it) — marks it delivered so it is not brought up again. Never call this for a background ' +
    'Job (use check_on_work/stop_working_on for those instead).',
  parameters: {
    type: 'object',
    properties: {
      notice_id: { type: 'string', description: 'The notice_id shown alongside the finding you just mentioned.' },
    },
    required: ['notice_id'],
  },
  async run(args) {
    const id = Number(args?.notice_id);
    if (!Number.isFinite(id)) return { ok: false, error: 'That notice_id was not valid.' };
    const entry = getById(id);
    if (!entry) return { ok: false, error: 'No notice with that id — it may already be resolved.' };
    if (entry.source !== 'heartbeat') return { ok: false, error: 'That was a background Job, not a notice — use check_on_work or stop_working_on instead.' };
    if (!entry.deliveredAt) markDelivered(id);
    return { ok: true };
  },
};

File: backend/jarvis/tools/acknowledge_notice.py
"""Marking a waiting notice as actually delivered.

A notice is shown to the model on a turn the user started, but is NOT marked
delivered by being shown: a turn that fails before the model replies would
otherwise lose it silently. This is the action that says it genuinely reached
the person — which is why it is a tool call rather than a side effect of
rendering the prompt.

Only for notices with no other resolving action of their own. Background work
has `check_on_work` and `stop_working_on`, which resolve their rows by doing
something about them.
"""

from __future__ import annotations

from typing import Any

from ..capabilities import CapabilitySpec, Risk


def _run(notice_id: int | str = 0) -> dict[str, Any]:
    from ..heartbeat import outbox

    try:
        entry_id = int(notice_id)
    except (TypeError, ValueError):
        return {"ok": False, "error": "That is not a notice number."}

    entry = outbox.get(entry_id)
    if entry is None:
        return {"ok": False, "error": "There is no notice with that number."}
    if entry["deliveredAt"]:
        return {"ok": True, "alreadyDone": True,
                "note": "That one was already dealt with — don't mention it again."}
    if entry["source"] == "job":
        # Its own actions resolve it by doing something about it; marking it
        # delivered here would drop it while the job is still waiting.
        return {"ok": False,
                "error": "That one is about background work — use check_on_work or "
                         "stop_working_on instead."}

    outbox.mark_delivered(entry_id)
    return {"ok": True, "note": "Noted as passed on. Don't mention it again unless asked."}


SPEC = CapabilitySpec(
    id="builtin.acknowledge_notice",
    name="acknowledge_notice",
    description=("Mark a waiting notice as passed on, once you have actually mentioned it to "
                 "the user. Call it with the notice number you were shown. Only for notices — "
                 "anything about background work is resolved with check_on_work or "
                 "stop_working_on instead."),
    input_schema={"type": "object", "properties": {
        "notice_id": {"type": "integer", "description": "The number shown with the notice."}},
        "required": ["notice_id"]},
    risk=Risk.LOW,
    handler=_run,
    timeout_s=10.0,
    tags=frozenset({"core", "meta"}),
)
